### PR TITLE
friend list

### PR DESCRIPTION
**Pull request by ThuyenPVT**
friend list : https://gyazo.com/31facdf1423ea4bc69e4dcdaccf23e52

### CODE REVIEW

File: lib/data/source/remote/account_remote_datasource.dart [hunk 1]
```diff
@@ -10,7 +10,16 @@ abstract class AccountRemoteDataSource {
     String imageURL,
   );
 
+  Future<void> sendFriendRequest(
+    String currentID,
+    recipientID,
```
Comment (chuongdevJC) [inline lib/data/source/remote/account_remote_datasource.dart]: why this params have no type value?

File: lib/data/source/remote/account_remote_datasource.dart [hunk 2]
```diff
@@ -42,4 +52,28 @@ class AccountRemoteDataSourceImpl implements AccountRemoteDataSource {
     final _snapshot = await _ref.get();
     return _snapshot.docs.map((doc) => Account.fromFireStore(doc)).toList();
   }
+
+  @override
+  List<QueryDocumentSnapshot> getAllUsers(
+      AsyncSnapshot<QuerySnapshot> snapshot) {
+    if (snapshot.data != null) {
```
Comment (chuongdevJC) [inline lib/data/source/remote/account_remote_datasource.dart]: please give a return

File: lib/data/source/remote/account_remote_datasource.dart [hunk 3]
```diff
@@ -42,4 +52,28 @@ class AccountRemoteDataSourceImpl implements AccountRemoteDataSource {
...
+  }
+
+  @override
+  Future<void> sendFriendRequest(
+    String currentID,
+    recipientID,
+    String name,
+    bool pending,
+  ) async {
+    try {
```
Comment (chuongdevJC) [inline lib/data/source/remote/account_remote_datasource.dart:54]: how to know if this request error?

File: lib/pages/friend_list/friend_list_screen.dart [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import 'package:structure_flutter/core/resource/text_style.dart';
+import 'package:structure_flutter/di/injection.dart';
+import 'package:structure_flutter/pages/friend_list/widgets/avatar.dart';
+import 'package:structure_flutter/repositories/account_repository.dart';
+import 'package:structure_flutter/widgets/loading_widget.dart';
+import 'package:structure_flutter/widgets/snackbar_widget.dart';
+
+import 'widgets/followers.dart';
+
+class FriendListScreen extends StatefulWidget {
```
Comment (chuongdevJC) [inline lib/pages/friend_list/friend_list_screen.dart]: FriendListScreen -> FriendListPage. all of widgets is page will named ...Page

File: lib/pages/friend_list/friend_list_screen.dart [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  final _accountRepository = getIt<AccountRepository>();
+
+  final _snackBar = getIt<SnackBarWidget>();
+
+  User get user => widget.user;
+
+  @override
+  Widget build(BuildContext context) {
+    //Validation data exist or not !
+    return StreamBuilder<QuerySnapshot>(
```
Comment (chuongdevJC) [inline lib/pages/friend_list/friend_list_screen.dart]: try to use bloC pattern

File: lib/data/source/remote/account_remote_datasource.dart (hunk 3 above)
Comment (ThuyenPVT) [reply lib/data/source/remote/account_remote_datasource.dart:54]: i think it's hard to get the request error in this case !

File: lib/pages/friend_list/friend_list_screen.dart (hunk 5 above)
Comment (ThuyenPVT) [reply lib/pages/friend_list/friend_list_screen.dart]: I already using _bloc_ in this case but i think the _stream_ is the pefect solution to solve this problem

File: lib/bloc/blocs/load_data_bloc.dart [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:structure_flutter/bloc/events/load_data_event.dart';
+import 'package:structure_flutter/bloc/states/load_data_state.dart';
+import 'package:structure_flutter/di/injection.dart';
+import 'package:structure_flutter/repositories/account_repository.dart';
+
+class LoadDataBloc extends Bloc<LoadDataEvent, LoadDataState> {
```
Comment (truyenjc) [inline lib/bloc/blocs/load_data_bloc.dart]: name is not enough meaning

File: lib/data/source/remote/account_remote_datasource.dart [hunk 7]
```diff
@@ -10,13 +11,32 @@ abstract class AccountRemoteDataSource {
...
+    bool pending,
+  });
+
+  Future<List<Account>> getUsersByName(String searchName);
+
+  List<QueryDocumentSnapshot> getAllUsers(
+    AsyncSnapshot<QuerySnapshot> snapshot,
+  );
+
+  Future<List<Account>> parseToObject();
```
Comment (truyenjc) [inline lib/data/source/remote/account_remote_datasource.dart]: like comment below
https://github.com/chuongdevJC/Structure_Flutter/pull/10#discussion_r518560085

File: lib/data/source/remote/account_remote_datasource.dart [hunk 8]
```diff
@@ -35,11 +56,36 @@ class AccountRemoteDataSourceImpl implements AccountRemoteDataSource {
...
     var _ref = _userCollection
         .where("name", isGreaterThanOrEqualTo: searchName)
         .where("name", isLessThan: searchName + 'z');
     final _snapshot = await _ref.get();
     return _snapshot.docs.map((doc) => Account.fromFireStore(doc)).toList();
   }
+
+  @override
+  List<QueryDocumentSnapshot> getAllUsers(
+      AsyncSnapshot<QuerySnapshot> snapshot) {
```
Comment (truyenjc) [inline lib/data/source/remote/account_remote_datasource.dart]: why does this function need to return ```QueryDocumentSnapshot```?

File: lib/pages/authen/register/widgets/heading.dart [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import 'package:flutter/material.dart';
+import 'package:structure_flutter/core/resource/text_style.dart';
+class Heading extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+      return Container(
```
Comment (truyenjc) [inline lib/pages/authen/register/widgets/heading.dart]: remove ```Container``` if its not necessary

File: lib/pages/friend_list/friend_list_page.dart [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+import 'package:structure_flutter/core/resource/assets_images.dart';
+import 'package:structure_flutter/data/entities/account.dart';
+import 'package:structure_flutter/di/injection.dart';
+import 'package:structure_flutter/repositories/account_repository.dart';
+import 'package:structure_flutter/widgets/loading_widget.dart';
+import 'package:structure_flutter/widgets/snackbar_widget.dart';
+
+import 'widgets/friend_profile.dart';
+
+class FriendListScreen extends StatefulWidget {
```
Comment (truyenjc) [inline lib/pages/friend_list/friend_list_page.dart]: rename -> ```FriendListPage``` to sync with others page

File: lib/pages/friend_list/friend_list_page.dart [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        return Loading();
+      },
+    );
+  }
+
+  String convertTimeStampToHour(Map<String, dynamic> data) {
+    int timestamp = data['lastSeen'].seconds;
+    var date = DateTime.fromMillisecondsSinceEpoch(timestamp);
+    return DateFormat.jm().format(date);
+  }
```
Comment (truyenjc) [inline lib/pages/friend_list/friend_list_page.dart]: extract to ```DateTimeUtils```

**Review by truyenjc — APPROVED** [review]
LGTM!!! (y)

**Pull request merged**